libavformat/rtcenc.c

```diff
@@ -0,0 +1,1548 @@
+
+    return 0;
+}
+
+static int rtc_write_header(AVFormatContext *s)
+{
+    int ret;
+
+    if ((ret = ice_handshake(s)) < 0)
+        return ret;
+
+#if CONFIG_OPENSSL
+    if ((ret = openssl_dtls_handshake(s)) < 0)
+        return ret;
+#else
+    av_log(s, AV_LOG_ERROR, "DTLS is not supported, please enable openssl\n");
+    return AVERROR(ENOSYS);
+#endif
+
+    if ((ret = setup_srtp(s)) < 0)
+        return ret;
+
+    if ((ret = create_rtp_muxer(s)) < 0)
+        return ret;
+
+    return ret;
+}
+
+static int rtc_write_packet(AVFormatContext *s, AVPacket *pkt)
+{
+    int ret, size, is_idr, i;
+    AVPacket* extra = NULL;
+    uint8_t *p;
+    RTCContext *rtc = s->priv_data;
+    AVStream *st = s->streams[pkt->stream_index];
+    AVFormatContext *rtp_ctx = st->priv_data;
+
+    /* TODO: Send binding request every 1s as WebRTC heartbeat. */
+
+    /* For audio OPUS stream, correct the timestamp. */
+    if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO) {
+        pkt->dts = pkt->pts = rtc->audio_jitter_base;
+        rtc->audio_jitter_base += 960;
+    }
+
+    /* Insert a packet with SPS/PPS before each IDR frame. */
+    is_idr = (pkt->flags & AV_PKT_FLAG_KEY) && st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO;
+    if (is_idr && st->codecpar->extradata) {
+        extra = av_packet_alloc();
+        size = !rtc->avc_nal_length_size ? st->codecpar->extradata_size :
+                rtc->avc_nal_length_size * 2 + rtc->avc_sps_size + rtc->avc_pps_size;
+        ret = av_new_packet(extra, size);
+        if (ret < 0) {
+            av_log(s, AV_LOG_ERROR, "Failed to allocate extra packet\n");
+            return ret;
+        }
+
+        /* Encode SPS/PPS in annexb format. */
+        if (!rtc->avc_nal_length_size) {
+            memcpy(extra->data, st->codecpar->extradata, size);
+        } else {
+            /* Encode SPS/PPS in ISOM format. */
+            p = extra->data;
+            for (i = 0; i < rtc->avc_nal_length_size; i++) {
+                *p++ = rtc->avc_sps_size >> (8 * (rtc->avc_nal_length_size - i - 1));
+            }
+            memcpy(p, rtc->avc_sps, rtc->avc_sps_size);
+            p += rtc->avc_sps_size;
+
+            /* Encode PPS in ISOM format. */
+            for (i = 0; i < rtc->avc_nal_length_size; i++) {
+                *p++ = rtc->avc_pps_size >> (8 * (rtc->avc_nal_length_size - i - 1));
+            }
+            memcpy(p, rtc->avc_pps, rtc->avc_pps_size);
+            p += rtc->avc_pps_size;
+        }
+
+        /* Setup packet and feed it to chain. */
+        extra->pts = pkt->pts;
+        extra->dts = pkt->dts;
+        extra->stream_index = pkt->stream_index;
+        extra->time_base = pkt->time_base;
+
+        ret = ff_write_chained(rtp_ctx, 0, extra, s, 0);
+        if (ret < 0)
+            goto end;
+    }
+
+    ret = ff_write_chained(rtp_ctx, 0, pkt, s, 0);
+
+end:
+    av_packet_free(&extra);
+    return ret;
+}
+
+static int rtc_write_trailer(AVFormatContext *s)
+{
+    return 0;
+}
```

Comment:
Will this function be used?

Reply:
Got it.

In `rtc_write_trailer`, I request server to close the stream, by sending a HTTP DELETE request. This is also defined in WHIP protocol:

```text
   To explicitly terminate a session, the WHIP client MUST perform an
   HTTP DELETE request to the resource URL returned in the Location
   header field of the initial HTTP POST.  Upon receiving the HTTP
   DELETE request, the WHIP resource will be removed and the resources
   freed on the Media Server, terminating the ICE and DTLS sessions.
```

See https://github.com/ossrs/ffmpeg-webrtc/pull/1/commits/4d564c829139bb816d79eb9c87ff1a7914093d4b